src/target/target.h

```diff
@@ -210,6 +210,9 @@ struct target {
 
 	/* The semihosting information, extracted from the target. */
 	struct semihosting *semihosting;
+#if BUILD_RISCV_ARTY_BSCAN == 1
+	int bscan_tunnel_ir_width;  /* if zero, then tunneling is not present/active */
```

Comment:
Since this is only implemented for RISC-V, this option should be in riscv_info_t, or possibly dm013_info_t if we don't expect to ever need this for 0.11.

Reply:
I'll try the 013 only option, thanks.

Reply:
Tim, the presence of this feature isn't something that can be dynamically queried, at least currently.  It's something that is declared in the startup script, as an attribute of the "target configure" command.  I guess that was the reason that putting it in struct target seemed natural, because the parsing of "target configure" happens in target.c.   If you think it is better to move the data to dm013_info_t, would declaring the data in "target configure" command still be the right thing, or should I declare it in some other way within the config file?  Some sort of risc-v specific initialization command?

Reply:
Or maybe if the IR width is declared as 6 within the "jtag newtap", is that enough to assume that a tunnel should be used?  The Arty TAP is 6 bits wide, and is declared that way in the newtap.  If a 6-bit wide rather than 5-bit wide IR is enough for riscv*.c logic to assume a BSCAN tunnel, maybe then that would be a possibility that removes the need for any code/data members for this in target.c, and removes the need for a parameter in "target configure"

Reply:
Since this feature is RISC-V-specific, it makes a lot of sense to configure it with a riscv command. See e.g. https://github.com/riscv/riscv-openocd/blob/riscv/src/target/riscv/riscv.c#L1830.

I think all those commands just do global configuration, but I think that will be OK. It would break if somebody ever has a JTAG chain consisting of an Arty board and a regular RISC-V TAP, but that sounds quite unlikely.

Relying on the IR width is more fragile.

Reply:
On the most recent commit, I reworked it to take the global configuration approach.  I did try adding the "use bscan" flag to dm013_info_t, rather than global state in riscv.c, but a problem with that is that the dm013_info_t struct hasn't been allocated yet at the first point at which we need to know whether or not to use BSCAN tunnel.  The code compiled, but at runtime there was a NULL pointer dereference segfault; an active struct target instance already exists at that point of OpenOCD initialization but the target->arch_info hasn't yet been assigned a non-NULL value.  There's probably a different way around that, but right now I have the "use bscan" as global within riscv.c.